### PR TITLE
Fix IPv6 address support in OUTPUT_URL/BACKUP_URL

### PR DESCRIPTION
**Pull request by rmetrich**
#### Relax-and-Recover (ReaR) Pull Request Template

Please fill in the following items before submitting a new pull request:

##### Pull Request Details:

* Type: **Bug Fix** / **Enhancement**

* Impact:  **Normal**

* How was this pull request tested? with both NFS and SSHFS schemes

* Description of the changes in this pull request:

Currently there is no proper support for IPv6 addresses in `OUTPUT_URL`/`BACKUP_URL` properties:
- `OUTPUT_URL=nfs://2001:db8:ca2:6::101/root/REAR` is not recognized (and should not be)
- `OUTPUT_URL=nfs://[2001:db8:ca2:6::101]/root/REAR` fails because of bash substitutions occurring in various places
- `OUTPUT_URL="nfs://\[2001:db8:ca2:6::101\]/root/REAR"` works but is somehow cumbersome (double quotes + square brackets quoting).

This patch makes the following (proper) scheme work with both NFS and SSHFS:
~~~
OUTPUT_URL=nfs://[2001:db8:ca2:6::101]/root/REAR
OUTPUT_URL=sshfs://[2001:db8:ca2:6::101]/root/REAR
~~~

For this to occur, a lot of bash quoting had to be fixed.

### CODE REVIEW

**Comment by rmetrich** [comment]
Yes all this mess was due to quoting. So I did a pass on everything that was closely related.

**Review by jsmeix — APPROVED** [review]
I approve it because - as far as I see - there are
only quoting changes of the form
`$string` -> `"$string"` or
`string=$( COMMAND )` -> `string="$( COMMAND )"`
where quoting is OK
but no changes like
`for word in $string` -> `for word in "$string"`
where quoting would be wrong.

**Comment by jsmeix** [comment]
@rear/contributors 
could one of you - as time permits - please
also have a look here and approve it if it looks OK.

Provided it is OK I would like to merge it rather soon
in a few days to avoid that other merged code changes
at all the various code places could cause merge conflicts
for this pull request which @rmetrich would have to solve.

**Review by schlomo — APPROVED** [review]
It looks like the only thing that changed is the quoting? Thanks a lot!

**Comment by pcahyna** [comment]
I reviewed CI results: the Fedora 40 / Rawhide errors are known and unrelated. There was a fail on CentOS Stream 9 due to this:
```
WARNING: tar ended with return code 1 and below output (last 5 lines):
                                ---snip---
                                tar: /var/log/audit/audit.log: file changed as we read it
                                ----------
                              This means that files have been modified during the archiving
                              process. As a result the backup may not be completely consistent
                              or may not be a perfect copy of the system. Relax-and-Recover
                              will continue, however it is highly advisable to verify the
                              backup in order to be sure to safely recover this system.
```
This is a harmless issue unrelated to the PR and rerun fixed it. The check for warnings in the test should probably be less zealous.

Differential ShellCheck says:
ℹ️ No Fixes!

which indicates that we should probably increase the strictness of the syntax checks. This change is definitely a fix of the shell syntax that ShellCheck is able to recognize.

**Review by pcahyna — APPROVED** [review]
Impressive work, @rmetrich , thank you a lot!

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [inline usr/share/rear/lib/global-functions.sh:675]: For consistency, `$mountpoint` should be quoted also here.

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [inline usr/share/rear/lib/global-functions.sh:678]: and here.

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [reply usr/share/rear/lib/global-functions.sh:675]: The `http://$(url_host "$url")$(url_path "$url")` part should probably be quoted.

File: usr/share/rear/backup/DUPLICITY/default/980_unmount_duplicity_path.sh
Comment (pcahyna) [inline usr/share/rear/backup/DUPLICITY/default/980_unmount_duplicity_path.sh:8]: `$BUILD_DIR/outputfs` should be quoted as well for consistency with other changes

File: usr/share/rear/backup/NETFS/default/100_mount_NETFS_path.sh
Comment (pcahyna) [inline usr/share/rear/backup/NETFS/default/100_mount_NETFS_path.sh:5]: `$BUILD_DIR/outputfs` should be quoted as well for consistency with other changes

File: usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh
Comment (pcahyna) [inline usr/share/rear/output/PXE/default/810_create_pxelinux_cfg.sh:105]: `$BUILD_DIR/tftpbootfs` should be quoted as well for consistency with other changes

File: usr/share/rear/output/default/980_umount_output_dir.sh
Comment (pcahyna) [inline usr/share/rear/output/default/980_umount_output_dir.sh:11]: `$BUILD_DIR/outputfs` should be quoted as well for consistency with other changes

File: usr/share/rear/prep/USB/default/060_set_usb_device.sh
Comment (pcahyna) [inline usr/share/rear/prep/USB/default/060_set_usb_device.sh:33]: This looks like an added whitespace error - scheme will have a trailing space and the code will malfunction.

File: usr/share/rear/backup/NETFS/default/500_make_backup.sh
Comment (pcahyna) [inline usr/share/rear/backup/NETFS/default/500_make_backup.sh:295]: If you quote `$backuparchive`, you should quote `$(dirname "$backuparchive")` and `$(basename "$backuparchive")` as well, otherwise the added quoting does not have much effect.

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [inline usr/share/rear/lib/global-functions.sh:635]: For consistency, `$mountpoint` and the URL should be quoted also here.

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [inline usr/share/rear/lib/global-functions.sh:637]: For consistency, `$mountpoint` and the URL should be quoted also here.

File: usr/share/rear/lib/global-functions.sh
Comment (pcahyna) [inline usr/share/rear/lib/global-functions.sh:641]: For consistency, `$mountpoint` and the URL should be quoted also here.

**Review by pcahyna — CHANGES_REQUESTED** [review]
I spotted one possibly serious mistake in the changes - please review and fix.

https://github.com/rear/rear/pull/3242/files/a543e9d5d18fa6992038fa0ef73a6834a2231296#diff-a1cd629146844c0b2a7c61dba11903f293a344e9c71cba2ff86787d95a7228a2

I also identified places where even more quoting should be added for consistency with other changes, feel free to ignore these, as the changes here are already a big improvement.

File: usr/share/rear/output/default/980_umount_output_dir.sh
Comment (rmetrich) [reply usr/share/rear/output/default/980_umount_output_dir.sh:11]: I didn't quote everything indeed, because I wanted minimal changes
but I agree this can be quoted as well, even though we know it's not problematic since under control somehow

**Comment by jsmeix** [comment]
@rmetrich 
take your time to fix the one mistake.
I will not merge it today.
I will wait until you finished.

If you find the time I would much appeciate it
when you could add quoting where @pcahyna mentioned it
because now we have the detailed review of @pcahyna 
which would be lost when this pull request was merged
without the additional quoting as mentioned by @pcahyna 

@pcahyna 
as always your precise and accurate review is so helpful
in particular for our users who are hit by less bugs
and also for us who need to fix less bugs.
Thank you so much for it!

File: usr/share/rear/prep/USB/default/060_set_usb_device.sh
Comment (rmetrich) [reply usr/share/rear/prep/USB/default/060_set_usb_device.sh:33]: well spotted!

File: usr/share/rear/output/default/980_umount_output_dir.sh
Comment (pcahyna) [reply usr/share/rear/output/default/980_umount_output_dir.sh:11]: I see, I checked only lines that you touched in any case, to not increase the volume of lines changed.

I don't entirely agree with "under control" though, as users can set BUILD_DIR to something that contains spaces or other special characters (although this is unlikely and fixing it is not a priority).

**Comment by jsmeix** [comment]
@pcahyna 
regarding your
https://github.com/rear/rear/pull/3242#issuecomment-2162636120
```
... we should probably increase the strictness
of the syntax checks. This change is definitely a fix
of the shell syntax that ShellCheck is able to recognize
```

As far as I know our Differential ShellCheck
only ckecks the syntax of the actual changes.

In this case I agree to try to change the level
of our Differential ShellCheck syntax checks
to let it report also quoting problems.

Since my
https://github.com/rear/rear/pull/3242#issuecomment-2158560575
```
Let's wait and see when things like [...]
will become regular characters in system file names
or device node names and so on - then we need to add
missing bash quoting at many more zillions of places.
```
I understand that proper quoting basically everywhere
is needed to avoid that bash expansion - in particular
bash pathname expansion - changes string values in
unintended ways.

BUT:
Proper quoting of all string values everywhere
requires to no longer use strings for multiple values
in one (string) variable as in
```
for word in $string ; do
    CODE that deals with $word
done
```
but to always use bash arrays for multiple values
in one (array) variable.
And this would be a huge cleanup effort.

**Comment by pcahyna** [comment]
@jsmeix 

> ... And this would be a huge cleanup effort.

Fortunately the check is differential, so it would not complain about existing code. But, when changing existing code which uses this idiom, one would need to do the cleanup anyway to avoid ShellCheck warnings on the changed lines, so some cleanup would be needed from time to time indeed.

**Comment by jsmeix** [comment]
@pcahyna 
and such small "by the way" cleanup parts step by step
as code is touched is the intended benefit to move our
code step by step towards porper quoting.

**Comment by pcahyna** [comment]
@rmetrich to be clear, I certainly don't want to request fixing quoting in more and more places if the changes you already did are sufficient for your purpose. That would be a huge effort. I merely suggested to take the opportunity to add more quoting to lines that you already touched, for consistency.

**Comment by jsmeix** [comment]
@rmetrich 
are you finished so it can be merged
or do need to do more changes?

**Comment by rmetrich** [comment]
I'm good, but would prefer that @pcahyna reviews again :-)

**Comment by jsmeix** [comment]
@rmetrich 
you are right, currently this pull request's state is
@pcahyna "requested changes"
so he needs to approve if now things are OK for him.

File: usr/share/rear/backup/BLOCKCLONE/default/400_copy_disk_struct_files.sh
Comment (pcahyna) [inline usr/share/rear/backup/BLOCKCLONE/default/400_copy_disk_struct_files.sh:11]: If you don't quote the result of the expansion here and also `$opath` below, quoting `$path` here will not have much effect in many cases (although that's not directly related to the IPv6 problem, because the shell metacharacters don't appear in the path part of the URL).

I don't see where is `path` getting defined anyway, so for the cases where `${path}` is returned on output, the code has not worked anyway, thus quoting is not that a big issue.

**Comment by pcahyna** [comment]
@rmetrich I reviewed again and found some unaddressed (low-priority) comments. Is that an oversight or you did not intend to address them? I also found another (low priority) issue that was present in the former code and not addressed by your changes.

**Comment by pcahyna** [comment]
@rmetrich thanks you for the new changes, I reviewed all of them and marked all comments that they address as resolved. Do you plan to address the rest (see the unresolved comments)? Some of them are in network protocols (CIFS, DAV) that would IMO be affected by IPv6 address problems just as NFS was. Other issues are not related and changes would be merely for consistency with other changes.

**Comment by rmetrich** [comment]
I consider the changes are all very invasive, I wouldn't add risk with fixing more stuff.

**Review by pcahyna — APPROVED** [review]

**Review by jsmeix — CHANGES_REQUESTED** [review]
Shit!
I messed it up with my latest commit here
where I overlooked the line with changed quoting.

**Comment by pcahyna** [comment]
@jsmeix , if you made a mistake, please reset your branch to the previous commit [4485326](https://github.com/rear/rear/pull/3242/commits/4485326d8ec2b62d04f757a0f5232a17caf52773) and force-push.

**Comment by jsmeix** [comment]
@rmetrich 
I am sorry I messed up your added quotation
with my too fast attempt to solve a merge conflict via
https://github.com/rear/rear/pull/3242/commits/690e03bb8a01ea5bfbdc08b033f0551efa85788c

I don't see how I could fix that myself here.

**Comment by pcahyna** [comment]
@jsmeix I will reset it then, ok?

**Comment by jsmeix** [comment]
@pcahyna 
yes please - of course it is OK for me.
Thank you!

**Comment by pcahyna** [comment]
And now my turn to make mistakes... I reset to a wrong commit, and I can't force-push anymore, damn.

@rmetrich can you please force-push again your commit 4485326 ?

**Comment by pcahyna** [comment]
... into your [rmetrich:ipv6](https://github.com/rmetrich/rear/tree/ipv6) branch.

**Comment by pcahyna** [comment]
Done, ufff. @jsmeix so now the branch is int he same state as when @rmetrich last touched it. What do you need to do now?

**Comment by pcahyna** [comment]
@jsmeix I will merge the PR and resolve the conflict if that's all that you wanted to do.

**Comment by rmetrich** [comment]
pushed again

**Comment by jsmeix** [comment]
@pcahyna @rmetrich 
thank you for fixing my mess!
I am sorry for that.

**Review by jsmeix — APPROVED** [review]

**Comment by pcahyna** [comment]
I opened two issues for the remaining unaddressed comments. Fixing them would need a good test coverage of the affected use cases though.

**Comment by jsmeix** [comment]
@pcahyna 
thank you for the two issues for the remaining missing quotings.
Because - as far as I know - no ReaR user was hit by it
(i.e. no ReaR user reported an issue here to us)
I think fixing the two issues is not required for ReaR 3.0
but it should be fixed (as time permits) for ReaR 3.1.
With the additional time we may even find a way how to
setup automated CI tests for the affected use cases.
I am in particular interested to learn how I could setup
automated CI tests for various SUSE specific standard systems,
cf. what I had manually tested for SLES12 and SLES15 in
https://github.com/rear/rear/wiki/Test-Matrix-rear-2.6#sles
But this is something for after ReaR 3.0 was released.

**Comment by pcahyna** [comment]
@jsmeix 

> @pcahyna thank you for the two issues for the remaining missing quotings.
> 
> Because - as far as I know - no ReaR user was hit by it (i.e. no ReaR user reported an issue here to us) I think fixing the two issues is not required for ReaR 3.0 but it should be fixed (as time permits) for ReaR 3.1.
> 
> With the additional time we may even find a way how to setup automated CI tests for the affected use cases.

Yes, I need to make the current test more generic though first, so that it is easier to add variations of it without having to copy it as a whole and modify it.

OTOH, for testing these bugs it may be helpful to add tests that are not full end-to-end tests (backup & recovery), but just backup or mkrescue tests, as the bugs will show up already at backup time. Such tests are much simpler and will not need to reuse the current Anton's and Lukas's backup&recover script. Of course this is not 100%, as the wrong quoting may be fixed in "mkbackup" scripts but left in "recover" scripts, and such a bug would be especially nasty from the user's POV.

> I am in particular interested to learn how I could setup automated CI tests for various SUSE specific standard systems, cf. what I had manually tested for SLES12 and SLES15 in https://github.com/rear/rear/wiki/Test-Matrix-rear-2.6#sles But this is something for after ReaR 3.0 was released.

Do you want full backup & recovery tests, or just mkbackup tests? The latter are much easier. Do you want a CI test here (on GitHub, to be triggered on every PR), or internal to SUSE? Our experience with Anton has been that the difficult part of public testing is finding a public infrastructure that has the features that you need. E.g. GitHub actions do not support rebooting the machine being tested. Testing Farm does, but you have only CentOS (Stream) and Fedora systems available in TF. Perhaps the support for other distros will appear eventually (I know there is already some support for Ubuntu and Debian, but not for VMs yet). Or maybe SUSE sponsors a different public infrastructure for such purposes?

> @pcahyna if you agree I set the milestone to "ReaR v3.1" for those two issues.

Sure, go ahead.

**Comment by pcahyna** [comment]
> Because I would like to automate what I had manually tested for SLES12 and SLES15 in https://github.com/rear/rear/wiki/Test-Matrix-rear-2.6#sles I would need full "rear mkbackup" plus "rear recover" tests.
> 
> Regarding missing support for "rebooting the machine being tested":
> 
> My immediate offhanded and totally untested idea is to use 'kexec' to boot the rear recovery system, cf. "Launching the ReaR recovery system via kexec" in https://en.opensuse.org/SDB:Disaster_Recovery and after "rear recover" use 'kexec' again to boot the recreated system, cf. #2186
> 
> Because 'kexec' should leave the disk as is, it should be possible to store what we like on disk in the old system before 'kexec' and have that available in the new system after 'kexec'. So when we use 'kexec' to boot the recreated system we could add some testing scripts that are run at the end when the recreated system started up.

Not sure whether this will work, depends on the reason why reboot is not possible - many test harnesses do not like to be killed and started again - the test script needs to run uninterrupted from the start of the test till the end. (The test harness we use in Testing Farm does not suffer from this limitation, we can kill the script, even the whole machine, and start it again, test will continue from the beginning.)

> Regarding "finding a public infrastructure" that can run SLES12 or SLES15:
> 
> My wishful thinking is that I could provide a manually "ready-to-use" prepaded '*.qcow2' image that can be booted by some public infrastructure in some QEMU/KVM virtual machine.

This would work, but you need to find an infrastructure that allows running VMs - as the test machines are usually VMs themselves, this requires nested virtualization working and enough resources. It looks like in GitHub Actions this is supported now but only on larger runners (not default): https://github.blog/changelog/2023-02-23-hardware-accelerated-android-virtualization-on-actions-windows-and-linux-larger-hosted-runners/

We can try if Testing Farm VMs support that.

**Comment by jsmeix** [comment]
@pcahyna 
thank you for the background information.
Now I can a little bit better imagine
what the root of the various problems is.
Unsurprisingly RFC 1925 item 6a comes to mind ;-)

Yes, I think that an infrastructure that allows
to run VMs directly (e.g. by providing a '*.qcow2' image)
could be the most simple and straightforward way.

As time permits I will ask at (open)SUSE
if we perhaps already have such an infrastructure.
E.g. currently I know basically nothing about openQA
http://open.qa/
which is what (open)SUSE uses for testing.

**Comment by jsmeix** [comment]
I was more thinking about possibly "misusing" openQA
but I it would not surprise me if openQA does not provide
what I would like to have - nevertheless I will ask,
perhaps someone at SUSE knows an infrastructure
as I would like to have.

But I would like to postpone that advanced CI testing
to after the ReaR 3.0 release (as time permits).

**Comment by pcahyna** [comment]
@jsmeix yes, certainly try asking around if there is an established way to test on SUSE distros on public infrastructure (just like Testing Farm is for Fedora-derived distros). Agree to keep it until after 3.0. I will also investigate about support for spawning your VMs inside Testing Farm VMs.

**Pull request merged**
